### ISSUE TITLE
Further improvements in Pull and Status calculation

### ISSUE THREAD

**Issue by sbcgua** [abapGit/abapGit#6107]
This may be one or several bugs or improvements to be done. Still investigating it ... 
And for the context: it's ZIP importing, but I'm sure the same will be with online

Let's suppose I have an object, I moved it to another package (1 to 2) and changed. Here is what I get:
![image](https://user-images.githubusercontent.com/15635498/221931967-dbd4695b-5a8f-444b-adaf-105acf1dfc23.png)
And also
![image](https://user-images.githubusercontent.com/15635498/221932068-4b18c7bb-61d2-414f-9644-1061abaf74e8.png)

Expected vs reality:
1) File status is misleading: the prog 2 is added remotely, not deleted locally. Also the change is not indicated
2) "Multiple files with same filename, zdesertest1.prog.abap" message is excessive - it is not so, it is because of package reassignment
3) The State in pull dialog is empty (and it is not the bug of pull dialog but of `object_checks->warning_overwrite_find`)
4) Last but not least: the action is wrong. It is not package reassignment but also a change. The change is hidden here and this is bad, it leads to wrong overview. The status and reassignment should be reflected independently (different actions ?)

More on point 4: package reassignments make a mess in complex deployments. It would be nice to be able to run reassignment first (maybe even separately from deserialization) to see a clearer status after it. 

Also in additional to the #6090 I this the "Select category" should have also options line "Reassignment only", "New only" ... Although it can be also filtered ... so just an idea.

Will have a look on these

**Comment by larshp** [comment]
we still dont have enough information to determine if it was removed/added/updated in remote/local...

**Comment by sbcgua** [comment]
For this particular case I have a suspect: the `ZCL_ABAPGIT_FILE_STATUS=>PROCESS_REMOTE` overwrites the status

```abap
      IF sy-subrc = 0 AND <ls_remote>-filename <> zcl_abapgit_filename_logic=>c_package_file.
        <ls_result>-match = abap_false.
        <ls_result>-lstate = zif_abapgit_definitions=>c_state-deleted.
        <ls_result>-rstate = zif_abapgit_definitions=>c_state-unchanged.
```

Before that, the status was correct. But ... still studying it

Also I have a feeling that there is an issue with checksums for ZIP. They are empty. Although I run recalculate explicitly. Strange. Also maybe if the check sums are empty they should be populated with local state (in `ZCL_ABAPGIT_FILE_STATUS=>CALCULATE_STATUS`). But also still studying it ...

**Comment by sbcgua** [comment]
And also the `packmove` flag seems to be not strict enough - it is set only for the case files are similar, but it's wrong. Or can be set for file and not for the other (XML and not abap)
```abap
* process_local
        " Check if same file exists in different location
        READ TABLE ct_remote ASSIGNING <ls_remote>
          WITH KEY file
          COMPONENTS filename = <ls_local>-file-filename.
        IF sy-subrc = 0 AND <ls_local>-file-sha1 = <ls_remote>-sha1. <<<<<<<<< WRONG
          <ls_result>-packmove = abap_true.
```

**Comment by sbcgua** [comment]
checksum and status calculation logic should probably be written somewhere in words. Or heavily commented. It's very difficult to judge (or remember ;) why this or that was done in the first place ...

E.g. the checksums. There is `ZCL_ABAPGIT_REPO_CHECKSUMS->BUILD_CHECKSUMS_FROM_FILES` (probably even written by me). It is used only for the rebuil_checksums. And it takes remote into account. But for example for ZIP it is not correct. Cause it just treats all files as new and does not calculate the CSs.

**Comment by larshp** [comment]
the checksums were initially added to determine if files had been over written locally since last pull

**Comment by sbcgua** [comment]
Yes, I know. My point is that it does not work accurate for offline repos at the moment. But I'll have a look on it.

**Comment by larshp** [comment]
IMHO we first need to fix the status, in order to fix the checksums

change status to only be
1: file only exists locally
2: file only exists remotely
3: files differ

then the checksums can/should be updated after pull, and only used to determine if files were changed locally to give overwrite errors when pulling

**Comment by mbtools** [comment]
If we reduce the status to one value (not separate local/remote), then there's no way to tell from which direction the change was made. Therefore, it won't be possible to decide which colors to use in diffs... red/green or green/red. 

The current system works well in the most common cases (like working within one branch). I can always tell, if it's a change that I didn't push (M-) or a change I didn't pull yet (-M) or if both sides where changed, which requires conflict resolution (MM). With just "different", we loose this essential info. 

We can change it. My only ask is that we spec out how it should work beforehand.

**Comment by larshp** [comment]
I never look at the "M" status, because I know its sometimes wrong. Personally I'd rather have correct bad information than smart incorrect information.

**Comment by sbcgua** [comment]
One more minor question on the way. Why `zcl_abapgit_file_status` belongs to OBJECTS_CORE ? It is more a repo logic. And as well as the `zcl_abapgit_objects_check` which uses it. Just for re-consideration

**Comment by sbcgua** [comment]
and `zcl_abapgit_item_state`

**Comment by larshp** [comment]
🤷‍♂️ move it? 🚚

**Comment by mbtools** [comment]
`zcl_abapgit_objects->deserialize/_check` depend on the classes you mentioned and we tried to move as dependencies out of `$abapgit_objects` to subpackages. but maybe the logic of which objects will/should be deserialized should to be outside of the object logic. quite a bit of refactoring...

**Comment by sbcgua** [comment]
Any ideas where `SELECT SINGLE editflag FROM trnspace INTO ls_trnspace-editflag WHERE namespace = lv_namespace.` from `ZCL_ABAPGIT_FILE_STATUS=>CHECK_NAMESPACE` would fit ? In case of decoupling the direct select ? `zcl_abapgit_sap_package` ? Or a separate new class ? Nothing critical but I just thought that would be nice to get rid of direct select in file_status class

**Comment by larshp** [comment]
trnspace: just make a new class `zcl_abapgit_namespace`, there is probably more somewhere that can/should be moved to it

**Comment by sbcgua** [comment]
zcl_abapgit_namespace created in #6124

**Comment by sbcgua** [comment]
Followups from #6093:

> Actually one usability concern ... intuitively, I'd expect that "Select All" (and also Deselect All) is applied to visible records ... Although it is not a hard rule and also a matter of habit, but may be a source of confusion ...

certainly confusing ... to be fixed, maybe even remove the "partial" button. Or leave it for line selected by mouse only (the scenario from Marc)

> "OK" (the default) should be on the left @mbtools  @christianguenter2 

- pro: same place for main buttons independent of other buttons layout (mouse user focused scenario)
- con: key navigation scenario: ... used to mark stuff with arrow keys and Space/Enter and then to press CTRL+TAB, Enter to continue
- option: change combination for OK ?
- ...

**Comment by sbcgua** [comment]
Maybe Ctrl-S for continue ? Or F2 (and move SEL_CAT to F8) ? @christianguenter2 what do you think ?

![image](https://user-images.githubusercontent.com/15635498/223393910-9f62bfeb-8df9-488e-beb2-af65e31279d2.png)

**Comment by christianguenter2** [comment]
CTRL+S seems like a good choice.
